### PR TITLE
HPCC-11367 review and clean up use of spinlocks

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith please review

File: common/remote/hooks/libarchive/archive.cpp [hunk 1]
```diff
@@ -608,7 +608,7 @@ class CArchiveFileHook : public CInterface, implements IContainedFileHook
 
 extern ARCHIVEFILE_API void installFileHook()
 {
-    SpinBlock b(*lock); // Probably overkill!
+    CriticalBlock b(*lock); // Probably overkill!
```
Comment (jakesmith) [inline common/remote/hooks/libarchive/archive.cpp]: Doesn't really matter what kind of lock it is here I suspect.
But we should default to use a CS if expect low contention/cost.
Is that your assertion?

File: roxie/ccd/ccdqueue.cpp [hunk 2]
```diff
@@ -43,8 +43,8 @@ unsigned numSlaves[MAX_CLUSTER_SIZE];
 unsigned replicationLevel[MAX_CLUSTER_SIZE];
 unsigned IBYTIDelays[MAX_CLUSTER_SIZE]; // MORE: this will cover only 2 slaves per channel, change to cover all. 
 
-SpinLock suspendCrit;
-bool suspendedChannels[MAX_CLUSTER_SIZE];
+static SpinLock suspendCrit; // MORE: Could remove this, and replace the following with an atomic boolean array.
+static bool suspendedChannels[MAX_CLUSTER_SIZE];
```
Comment (jakesmith) [inline roxie/ccd/ccdqueue.cpp]: curious why you didn't?

File: roxie/ccd/ccdsnmp.cpp [hunk 3]
```diff
@@ -852,7 +852,7 @@ class CQueryStatsAggregator : public CInterface, implements IQueryStatsAggregato
     CIArrayOf<QueryStatsAggregateRecord> aggregated; // stored with most recent first
     unsigned expirySeconds;  // time to keep exact info (rather than just aggregated)
     StringAttr queryName;
-    SpinLock lock;
+    SpinLock lock; // MORE: This could be held this for a while.  Is this significant?  Should it be a CriticalSection?
```
Comment (jakesmith) [inline roxie/ccd/ccdsnmp.cpp]: doubt massively significant this code is used, but does sound like a candidate to switch to CS to me.

**Comment by jakesmith** [comment]
@ghalliday - a couple of questions, but see no issues with the changes.

File: roxie/ccd/ccdsnmp.cpp [hunk 4]
```diff
@@ -908,7 +908,6 @@ class CQueryStatsAggregator : public CInterface, implements IQueryStatsAggregato
     CQueryStatsAggregator(const char *_queryName, unsigned _expirySeconds)
         : queryName(_queryName)
     {
-        SpinBlock b(queryStatsCrit);
```
Comment (richardkchapman) [inline roxie/ccd/ccdsnmp.cpp]: Why was this removed?

File: common/remote/hooks/libarchive/archive.cpp (hunk 1 above)
Comment (ghalliday) [inline common/remote/hooks/libarchive/archive.cpp]: My current position (subject to change!) is that we should only be using critical sections when the code is very time critical, and the work done in the critical section is as small as possible.

The spin locks have two problems:  On multiple socket systems under high contention they perform badly.  More significantly if the number of threads exceeds the number of cores they can perform terribly if the thread that holds the lock is descheduled.

So they have a relatively minor benefit if we can guarantee low contention and little work inside them, but a potentially bad worse-case behaviour.  Unless there is a clear benefit I think we should err on the side of critical sections.

File: roxie/ccd/ccdqueue.cpp (hunk 2 above)
Comment (ghalliday) [inline roxie/ccd/ccdqueue.cpp]: Because I wasn't sure how likely it was to be a significant issue.

@richardkchapman you understand that better than me.  What do you think?

File: roxie/ccd/ccdsnmp.cpp (hunk 4 above)
Comment (ghalliday) [inline roxie/ccd/ccdsnmp.cpp]: It was removed in error.  I'll add it back in again.

I had removed it because in general there is no point having a critical block inside a constructor.  However I see this was protecting adding the aggregator to the array so it is needed.  I'll also add a comment to indicate why it is needed.

**Comment by ghalliday** [comment]
@richardkchapman see comments above.  Fix released for constructor.

File: roxie/ccd/ccdqueue.cpp (hunk 2 above)
Comment (richardkchapman) [inline roxie/ccd/ccdqueue.cpp]: I don't believe anyone has ever used this capability, so not a big deal either way

File: common/remote/hooks/libarchive/archive.cpp (hunk 1 above)
Comment (richardkchapman) [inline common/remote/hooks/libarchive/archive.cpp]: Prior thinking was that spinlocks were faster than critsecs where contention was very unlikely. Did you mean critsec or spinlock in your first paragraph above?
